utils/utils.go

```diff
@@ -139,3 +152,65 @@ func GetRouteReference(r kong.Route) *kong.Route {
 	}
 	return route
 }
+
+// ParseKongVersion takes a version string from the Gateway and
+// turns it into a semver-compliant version to be used for
+// comparison across the code.
+func ParseKongVersion(version string) (semver.Version, error) {
+	v, err := CleanKongVersion(version)
+	if err != nil {
+		return semver.Version{}, err
+	}
+	return semver.ParseTolerant(v)
+}
+
+// ConfigFilesInDir traverses the directory rooted at dir and
+// returns all the files with a case-insensitive extension of `yml` or `yaml`.
+func ConfigFilesInDir(dir string) ([]string, error) {
+	var res []string
+	err := filepath.Walk(
+		dir,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if info.IsDir() {
+				return nil
+			}
+			switch strings.ToLower(filepath.Ext(path)) {
+			case ".yaml", ".yml", ".json":
+				res = append(res, path)
+			}
+			return nil
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("reading state directory: %w", err)
+	}
+	return res, nil
+}
+
+// CheckRoutePaths300AndAbove checks routes' paths format and prints out
+// a warning if regex-like patters without the '~' prefix are found.
+func CheckRoutePaths300AndAbove(route kong.Route) {
+	pathsWarnings := []string{}
+	for _, p := range route.Paths {
+		if strings.HasPrefix(*p, "~/") || !IsPathRegexLike(*p) {
+			continue
+		}
+		pathsWarnings = append(pathsWarnings, *p)
+	}
+	if len(pathsWarnings) > 0 {
+		cprint.UpdatePrintf(
+			"In Route '%s', a path with regular expression was detected.\n"+
+				"In Kong Gateway versions 3.0 and above, all paths that use regular expressions \n"+
+				"must be prefixed with a ~ character. Without the ~ prefix, regular expression \n"+
+				"based paths will not be matched and processed correctly. \n\n"+
+				"Please run the following command to upgrade your config: \n\n"+
+				"deck convert --from kong-gateway-2.x --to kong-gateway-3.x "+
+				"--input-file <config-file> --output-file <new-config-file>\n\n"+
+				"Please refer to the following document for further details:\n"+
+				"https://docs.konghq.com/deck/latest/3.0-upgrade.\n\n",
+			*route.ID)
+	}
```

Comment:
This is really strange, If there a 10k paths with regex routes, you are going to print the warning 10k times!? That is almost draconian.
I'd recommend printing the count of routes (that violate the expected prefix requirement) and the first 10 route names/IDs that have such problems.

Reply:
https://github.com/Kong/deck/pull/753/commits/abcdf7769376f9864858925c951aa31eb89323cb